### DISCUSSION TITLE
Parent model is not injected related children

### DISCUSSION THREAD

**Discussion by josephwynn-sc** [laravel/framework#49087] [Ideas]
### Laravel Version

10.33.0

### PHP Version

8.1.25

### Database Driver & Version

MariaDB 10.11

### Description

Eloquent does not seem to inject the parent model into related child models. This causes Eloquent to re-fetch the parent from the database whenever the child accesses its parent relation. This occurs regardless of using lazy or eager fetching. Ideally the parent model would be injected so that no further queries are made.

### Steps To Reproduce

1. Create two models with a one-to-many relationship, for example `Owner` (one) and `Item` (many)
2. Fetch a single owner object. Note one query is made to the database.
3. Access the items belonging to that owner. Note that another query is made.
4. Now access one of those items' owner property. Note that another query is made, even though it should already have access to the owner that caused it to be fetched.

```php
<?php
// 2 queries are made here
$owner = Owner::with(['items'])->first();

// 1 unnecessary query is made here 
$owner->items[0]->owner;

// Even this will make 3 queries, when it should need 2 at the most
Owner::with(['items', 'items.owner'])->first();

class Owner extends Model
{
    public function items()
    {
        return $this->hasMany(Item::class);
    }
}

class Item extends Model
{
    public function owner()
    {
        return $this->belongsTo(Owner::class);
    }
}
```

**Comment by josephwynn-sc** [comment]
If anyone else has this same issue, we resolved it in our API resources by instantiating child resource collections with a custom method that sets the parent relation on each of the children:

```php
<?php

use Illuminate\Database\Eloquent\Collection;
use Illuminate\Http\Resources\Json\AnonymousResourceCollection;

trait ExtensibleCollectionTrait
{
    public static function collectionWith(Collection $collection, array $properties): AnonymousResourceCollection
    {
        return static::collection($collection->map(function ($item) use ($properties) {
            foreach ($properties as $prop => $val) {
                $item->{$prop} = $val;
            }

            return $item;
        }));
    }
}
```

We added that trait to our child resource, and changed our API resources like so:

```diff
<?php

use Illuminate\Http\Resources\Json\JsonResource;

class OwnerResource extends JsonResource
{
    public function toArray($request)
    {
        return [
            'id' => $this->id,
-           'items' => ItemResource::collection($this->items),
+           'items' => ItemResource::collectionWith($this->items, [
+               'owner' => $this->resource,
+           ]),
        ];
    }
}
```

Now the `$owner` property is already set on each of the children, and Eloquent no longer fetches each of them from the database. For one of our endpoints this reduced the number of queries from about 800 to 15, because one of the child models was fetching other relations from the parent.

**Comment by ghost** [comment]
I see a memory issue here. Maybe I'm wrong, but if you have 1000 children, you will be storing in each of them the same parent. So if your case has small amount of children your solution suits your case, but for general purpose... What do you think?


You can test like:

    $owner = Owner::with(['items'])->first();
    //  modify owner here
    $owner->items[0]->owner;
    //  if the item owner is altered the memory issue does not exist.

Also if this would be the default behaviour, then when responding in json on get /owners?with=items you would get the owner with a property items: [] and on each item the owner again.

**Reply by valerio-bozzolan** [reply]
I don't see a memory issue in the workaround proposed by @osephwynn-sc because their proposed workaround is about storing 1 *additional* object reference for each entry, and this is very RAM friendly, since we are adding a reference to the very same parent object, and not a copy of that (like it happens as default with the current limited `with()` logic) - without even mentioning that this saves N queries - as already said.

**Comment by valerio-bozzolan** [comment]
I think that Laravel could and should manage this speculation under its hood, with just `with()` as you mentioned, without writing custom collection fillers.

I'm inclined to suggest to take your question and elevate as issue, as feature request.

**Reply by valerio-bozzolan** [reply]
P.S. maybe `chaperone()` does the trick since Laravel 11?

https://laravel.com/docs/11.x/eloquent-relationships
